### PR TITLE
fix(lib/grandpa): check equivocatory votes count

### LINKED ISSUE #2401: GRANDPA Does Not Check Multiplicity of Equivocatory Votes (GSR-07)
An equivocatory vote occurs when a validator has two or more votes at the same stage during a single round.

The specifications state that a valid justification must not contain more than two equivocatory votes from each voter.

The Gossamer function `getEquivocatoryVoters()` does not put a restriction on the number of equivocatory votes from each voter.

The impact is that commits that may be rejected in Substrate but will be accepted in Gossamer. Furthermore, it is an inefficient use of space and bandwidth to include more than two equivocatory votes.

The issue can be mitigated by rejecting commits with more than two equivocations per user.

This can be implemented by having `getEquivocatoryVoters()` return error if any authority has more than two equivocations as seen below.

```go
func getEquivocatoryVoters(votes []AuthData) (map[ed25519.PublicKeyBytes]struct{}, error) {
	eqvVoters := make(map[ed25519.PublicKeyBytes]struct{})
	voters := make(map[ed25519.PublicKeyBytes]int, len(votes))
	for _, v := range votes {
		voters[v.AuthorityID]++
		if voters[v.AuthorityID] > 2 {
			return nil, errors.New("Invalid Multiplicity for Voter")
		}
		if voters[v.AuthorityID] == 2 {
			eqvVoters[v.AuthorityID] = struct{}{}
		}
	}
	return eqvVoters, nil
}
```

### PR DESCRIPTION
**Pull request by kishansagathiya**
## Changes
A valid justification can't have more than two equivocatory votes per
voter. Check the number of equivocatory votes per voter and throw an
error if more than two.

## Tests
`lib/grandpa` tests should be enough

## Issues

Fixes #2401
## Primary Reviewer

@timwu20

### CODE REVIEW

File: lib/grandpa/message_handler_test.go [hunk 1]
```diff
@@ -649,7 +649,7 @@ func TestMessageHandler_VerifyBlockJustification_WithEquivocatoryVotes(t *testin
 
 	round := uint64(1)
 	number := uint32(1)
-	precommits := buildTestJustification(t, 20, round, setID, kr, precommit)
+	precommits := buildTestJustification(t, 18, round, setID, kr, precommit)
```
Comment (kishansagathiya) [inline lib/grandpa/message_handler_test.go:652]: there are 9 authorities. If we vote count is 20, there would be few authorities with more than 2 votes. More than two equivocatory votes make the justification invalid.

File: lib/grandpa/message_handler.go [hunk 2]
```diff
@@ -284,27 +284,32 @@ func (h *MessageHandler) verifyCatchUpResponseCompletability(prevote, precommit
 	return nil
 }
 
-func getEquivocatoryVoters(votes []AuthData) map[ed25519.PublicKeyBytes]struct{} {
+func getEquivocatoryVoters(votes []AuthData) (map[ed25519.PublicKeyBytes]struct{}, error) {
```
Comment (qdm12) [inline lib/grandpa/message_handler.go:287]: nit named return values would be nice

```suggestion
func getEquivocatoryVoters(votes []AuthData) (eqvVoters map[ed25519.PublicKeyBytes]struct{}, err error) {
```

File: lib/grandpa/message_handler.go [hunk 3]
```diff
@@ -284,27 +284,32 @@ func (h *MessageHandler) verifyCatchUpResponseCompletability(prevote, precommit
...
 	eqvVoters := make(map[ed25519.PublicKeyBytes]struct{})
 	voters := make(map[ed25519.PublicKeyBytes]int, len(votes))
 
 	for _, v := range votes {
 		voters[v.AuthorityID]++
+		if voters[v.AuthorityID] > 2 {
+			return nil, errInvalidMultiplicity
```
Comment (qdm12) [inline lib/grandpa/message_handler.go:294]: nit perhaps wrap some context information?

```suggestion
			return nil, fmt.Errorf("%w: authority id %x has %d votes", 
				errInvalidMultiplicity, v.AuthorityID, voters[v.AuthorityID])
```

File: lib/grandpa/message_handler.go [hunk 4]
```diff
@@ -284,27 +284,32 @@ func (h *MessageHandler) verifyCatchUpResponseCompletability(prevote, precommit
...
+		}
 
-		if voters[v.AuthorityID] > 1 {
+		if voters[v.AuthorityID] == 2 {
```
Comment (qdm12) [inline lib/grandpa/message_handler.go:297]: nit a switch could be nice:

```go
switch voters[v.AuthorityID] {
	case 1:
	case 2:
		eqvVoters[v.AuthorityID] = struct{}{}
	default:
		return nil, fmt.Errorf("%w: authority id %x has %d votes", 
			errInvalidMultiplicity, v.AuthorityID, voters[v.AuthorityID])
}
```

**Review by qdm12 — APPROVED** [review]
Good work! :+1:

File: lib/grandpa/message_handler.go (hunk 4 above)
Comment (timwu20) [reply lib/grandpa/message_handler.go:298]: I prefer a switch here as well.  Or at the very least an `else if` condition.

**Review by timwu20 — CHANGES_REQUESTED** [review]

**Review by EclesioMeloJunior — APPROVED** [review]
LGMT! just a small comment

**Review by timwu20 — APPROVED** [review]

**Pull request merged**